### PR TITLE
CLI

### LINKED ISSUE #29: Add tests for command line
ensure coverage of `bin/quence.js`

Might need to switch to commander for this to work right.  Might need to pull most of the work into `lib/cli.js`.

### PR DESCRIPTION
**Pull request by hildjj**
fixes #29

Switch from `optimist` to `commander` because a) optimist is deprecated and b) I've used commander a lot, and I like how it allows for plugging into tests.

Added support for "-" as input or output.

This is my stopping point for today.  There's still some testing of the new CLI that needs to go in.

@ekr, can you review this before I merge, please?  In particular, the option processing stuff changed a little to `-O foo=bar -O baz=12`, and numeric options are correctly coerced to type number now.

### CODE REVIEW

File: lib/driver.js [hunk 1]
```diff
@@ -9,19 +9,18 @@ export default class Driver {
...
-    }
+    this.argv.option?.forEach(opt => {
+      const [k, v] = opt.split('=').map(s => s.trim())
+      let value = v
+      if (v === undefined || v === 'true') {
+        value = true
+      } else if (v === 'false') {
+        value = false
+      } else if (typeof diag.props[k] === 'number') {
+        value = parseInt(v, 10)
```
Comment (hildjj) [inline lib/driver.js:20]: check for NaN and throw an exception

File: lib/cli.js [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+    .arguments('<FILE...>')
+    .option('-n, --nolink', 'do not add project link to output')
+    .addOption(
+      new Option('-o, --output <type>', 'output type')
+        .choices(VALID_OUTPUTS)
+        .default('pdf')
+    )
+    .option('--out <FILE>', 'output file name')
+    .option(
+      '-O, --option <key=value>',
```
Comment (hildjj) [inline lib/cli.js]: Change this to `-p, --property` for consistency with internal naming

File: test/cli.ava.js [hunk 3]
```diff
@@ -0,0 +1,74 @@
+import Store from './store.js'
```
Comment (hildjj) [inline test/cli.ava.js:1]: Add `execa`.  Execute the real CLI once, passing in a real stdin.  This is slower than testing through the API, but will allow us to check that the linkage to the lib script is still working.

File: package.json [hunk 4]
```diff
@@ -1,13 +1,13 @@
 {
   "name": "quence",
   "version": "1.1.1",
-  "main": "./lib/quence.js",
+  "main": "./lib/index.js",
```
Comment (hildjj) [inline package.json:4]: Fix up web subdir to match this change

File: lib/cli.js [hunk 5]
```diff
@@ -0,0 +1,101 @@
+import {Command, Option} from 'commander'
+import {SyntaxError, VALID_OUTPUTS, draw} from './index.js'
+import {name, version} from '../package.js'
+import fs from 'fs'
+import l4js from 'log4js'
+import path from 'path'
+
+const log = l4js.getLogger()
```
Comment (hildjj) [inline lib/cli.js:8]: I'm now questioning the utility of all of this logging.  Perhaps it should all be ripped out?  Add an issue to discuss.

File: lib/cli.js [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+    let data = ''
+    process.stdin.on('data', chunk => {
+      data += chunk
+    })
+    process.stdin.on('end', () => resolve(data))
+    process.stdin.on('error', reject)
+  })
+}
+
+async function readFile(fileName, opts, testing) {
```
Comment (hildjj) [inline lib/cli.js]: rename to `processFile`

File: lib/cli.js (hunk 2 above)
Comment (hildjj) [reply lib/cli.js]: Change README.md to match

File: lib/cli.js [hunk 7]
```diff
@@ -0,0 +1,101 @@
+import {Command, Option} from 'commander'
+import {SyntaxError, VALID_OUTPUTS, draw} from './index.js'
+import {name, version} from '../package.js'
+import fs from 'fs'
+import l4js from 'log4js'
+import path from 'path'
+
+const log = l4js.getLogger()
+
+function stdin(testing) {
```
Comment (hildjj) [inline lib/cli.js]: remove the `testing` param

File: test/cli.ava.js [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+
+test.before('Capture log events', () => {
+  l4js.configure({
+    appenders: {test: {type: 'recording'}},
+    categories: {default: {appenders: ['test'], level: 'WARN'}},
+  })
+})
+
+test.after('Reset logging', () => {
+  l4js.configure({
```
Comment (hildjj) [inline test/cli.ava.js:45]: What I want is either a `reset` method or the ability to read the current config so I can put it back.  Not having either leaves me questioning the log4js library.

File: package.json (hunk 4 above)
Comment (hildjj) [reply package.json:4]: not needed the way it was imported.

File: lib/cli.js (hunk 5 above)
Comment (hildjj) [reply lib/cli.js:8]: Added #31

**Comment by hildjj** [comment]
I feel pretty good about this now.  @ekr, I changed your option to `-p` and fixed up the docs to match.  We need to push to a release soon, because the docs on the website are now ahead of what people can install.

**Comment by ekr** [comment]
SG. I was going to review today but I see you merged, so do you need
anything from me?

**Comment by hildjj** [comment]
No, the main issues that needed your attention were the change in syntax to `-p foo -p bar=12`, and noticing the change to parsing numbers, which you would have found in testing.  If there are other tests you'd like to add, feel free in your next commit.

**Pull request merged**